pkg/protocol/block_dispatcher.go

```diff
@@ -0,0 +1,345 @@
+	processedWarpSyncRequests ds.Set[iotago.CommitmentID]
+
+	// shutdownEvent is a reactive event that is triggered when the BlockDispatcher instance is stopped.
+	shutdownEvent reactive.Event
+
+	// optWarpSyncWindowSize is the optional warp sync window size.
+	optWarpSyncWindowSize iotago.SlotIndex
+}
+
+// NewBlockDispatcher creates a new BlockDispatcher instance.
+func NewBlockDispatcher(protocol *Protocol, opts ...options.Option[BlockDispatcher]) *BlockDispatcher {
+	return options.Apply(&BlockDispatcher{
+		protocol:                  protocol,
+		dispatchWorkers:           protocol.Workers.CreatePool("BlockDispatcher.Dispatch"),
+		warpSyncWorkers:           protocol.Workers.CreatePool("BlockDispatcher.WarpSync", 1),
+		unsolidCommitmentBlocks:   buffer.NewUnsolidCommitmentBuffer[*types.Tuple[*model.Block, network.PeerID]](20, 100),
+		pendingWarpSyncRequests:   eventticker.New[iotago.SlotIndex, iotago.CommitmentID](eventticker.RetryInterval[iotago.SlotIndex, iotago.CommitmentID](WarpSyncRetryInterval)),
+		processedWarpSyncRequests: ds.NewSet[iotago.CommitmentID](),
+		shutdownEvent:             reactive.NewEvent(),
+	}, opts, func(b *BlockDispatcher) {
+		protocol.HookConstructed(b.initEngineMonitoring)
+		protocol.HookInitialized(b.initNetworkConnection)
+		protocol.HookStopped(b.shutdown)
+	})
+}
+
+// Dispatch dispatches the given block to the correct engine instance.
+func (b *BlockDispatcher) Dispatch(block *model.Block, src network.PeerID) error {
+	slotCommitment := b.protocol.ChainManager.LoadCommitmentOrRequestMissing(block.ProtocolBlock().SlotCommitmentID)
+	if !slotCommitment.SolidEvent().WasTriggered() {
+		if !b.unsolidCommitmentBlocks.Add(slotCommitment.ID(), types.NewTuple(block, src)) {
+			return ierrors.Errorf("failed to add block %s to unsolid commitment buffer", block.ID())
+		}
+
+		return ierrors.Errorf("failed to dispatch block %s: slot commitment %s is not solid", block.ID(), slotCommitment.ID())
+	}
+
+	matchingEngineFound := false
+	for _, engine := range []*engine.Engine{b.protocol.MainEngineInstance(), b.protocol.CandidateEngineInstance()} {
+		if engine != nil && (engine.ChainID() == slotCommitment.Chain().ForkingPoint.ID() || engine.BlockRequester.HasTicker(block.ID())) {
+			if !b.inWarpSyncRange(engine, block) {
+				engine.ProcessBlockFromPeer(block, src)
+			}
+
+			matchingEngineFound = true
+		}
+	}
+
+	if !matchingEngineFound {
+		return ierrors.Errorf("failed to dispatch block %s: no matching engine found", block.ID())
+	}
+
+	return nil
+}
+
+// initEngineMonitoring initializes the automatic monitoring of the engine instances.
+func (b *BlockDispatcher) initEngineMonitoring() {
+	b.monitorLatestEngineCommitment(b.protocol.mainEngine)
+
+	b.protocol.engineManager.OnEngineCreated(b.monitorLatestEngineCommitment)
+
+	b.protocol.Events.ChainManager.CommitmentPublished.Hook(func(chainCommitment *chainmanager.ChainCommitment) {
+		chainCommitment.SolidEvent().OnTrigger(func() {
```

Comment:
Can we add a comment about the flow here? This seems to be a key action why this whole thing works: once a commitment becomes solid in a chain, we know which engine it belongs to. Which in turn will allow us to advance the engine (and thus its created commitments/committed slots).